File: src/crypto/eth.ts
/**
 * Ethereum address.
 *
 * Validates the canonical 0x-prefixed 20-byte
 * hexadecimal address shape. The compact form is
 * lowercased so checksummed and non-checksummed
 * variants of the same address compare equal.
 */

import type {
  GlobalValidator,
  ValidateResult,
} from "../types";

import { keccak256 } from "#checksums/keccak";
import { clean } from "#util/clean";
import { err } from "#util/result";

const ETH_ADDRESS_LENGTH = 42;
const HEX_PREFIX_LENGTH = 2;

const compact = (value: string): string =>
  clean(value, " ").toLowerCase();

const asciiBytes = (value: string): Uint8Array => {
  const bytes = new Uint8Array(value.length);
  for (let i = 0; i < value.length; i += 1) {
    bytes[i] = value.charCodeAt(i);
  }
  return bytes;
};

const hexBytes = (bytes: Uint8Array): string => {
  let result = "";
  for (const byte of bytes) {
    result += byte.toString(16).padStart(2, "0");
  }
  return result;
};

const isMixedCase = (value: string): boolean =>
  /[a-f]/.test(value) && /[A-F]/.test(value);

const hasValidEip55Checksum = (
  address: string,
): boolean => {
  const body = address.slice(HEX_PREFIX_LENGTH);
  if (!isMixedCase(body)) return true;

  const lower = body.toLowerCase();
  const hash = hexBytes(keccak256(asciiBytes(lower)));

  for (let i = 0; i < body.length; i += 1) {
    const ch = body.charAt(i);
    if (/\d/.test(ch)) continue;

    const nibble = Number.parseInt(hash.charAt(i), 16);
    const expected =
      nibble >= 8 ? ch.toUpperCase() : ch.toLowerCase();
    if (ch !== expected) return false;
  }

  return true;
};

const validate = (value: string): ValidateResult => {
  const raw = clean(value, " ");
  const v = raw.toLowerCase();
  if (v.length !== ETH_ADDRESS_LENGTH) {
    return err(
      "INVALID_LENGTH",
      "Ethereum address must be 42 characters",
    );
  }
  if (!/^0x[0-9a-f]{40}$/i.test(raw)) {
    return err(
      "INVALID_FORMAT",
      "Ethereum address must be 0x followed by 40 hexadecimal characters",
    );
  }
  if (!hasValidEip55Checksum(raw)) {
    return err(
      "INVALID_CHECKSUM",
      "Ethereum address fails EIP-55 checksum validation",
    );
  }
  return { valid: true, compact: v };
};

const format = (value: string): string => compact(value);

const eth: GlobalValidator = {
  scope: "global",
  name: "Ethereum Address",
  localName: "Ethereum Address",
  abbreviation: "ETH",
  aliases: [
    "Ethereum address",
    "EVM address",
    "crypto wallet",
  ] as const,
  candidatePattern: "0x[0-9A-Fa-f]{40}",
  entityType: "any",
  examples: [
    "0xde709f2102306220921060314715629080e2fb77",
  ] as const,
  lengths: [ETH_ADDRESS_LENGTH] as const,
  compact,
  format,
  validate,
};

export default eth;
export { compact, format, hasValidEip55Checksum, validate };

File: crates/stdnum-core/src/validators/crypto/eth.rs
use crate::{
  compact_without, eth_has_valid_eip55_checksum,
  types::{
    EntityType, ValidationError, ValidationResult, Validator, ValidatorScope,
    ValidatorSpec,
  },
};

const ADDRESS_LENGTH: usize = 42;

pub static VALIDATOR: Validator = Validator::new(ValidatorSpec {
  id: "crypto.eth",
  name: "Ethereum Address",
  local_name: "Ethereum Address",
  abbreviation: "ETH",
  aliases: &["Ethereum address", "EVM address", "crypto wallet"],
  candidate_pattern: r"0x[0-9A-Fa-f]{40}",
  scope: ValidatorScope::Global,
  entity_type: EntityType::Any,
  source_url: None,
  lengths: &[ADDRESS_LENGTH],
  examples: &["0xde709f2102306220921060314715629080e2fb77"],
  compact,
  format,
  validate,
  generate: None,
  parse: None,
});

#[must_use]
pub fn compact(value: &str) -> String {
  compact_without(value, &[' ']).to_lowercase()
}

#[must_use]
pub fn format(value: &str) -> String {
  compact(value)
}

#[must_use]
pub fn has_valid_eip55_checksum(address: &str) -> bool {
  address.get(2..).is_some_and(eth_has_valid_eip55_checksum)
}

pub fn validate(value: &str) -> ValidationResult {
  let raw = compact_without(value, &[' ']);
  let compact = raw.to_lowercase();
  if compact.len() != ADDRESS_LENGTH {
    return Err(ValidationError::InvalidLength(
      "Ethereum address must be 42 characters",
    ));
  }
  let shape_valid = raw
    .get(..2)
    .is_some_and(|prefix| prefix.eq_ignore_ascii_case("0x"))
    && raw
      .get(2..)
      .is_some_and(|body| body.chars().all(|ch| ch.is_ascii_hexdigit()));
  if !shape_valid {
    return Err(ValidationError::InvalidFormat(
      "Ethereum address must be 0x followed by 40 hexadecimal characters",
    ));
  }
  if !has_valid_eip55_checksum(&raw) {
    return Err(ValidationError::InvalidChecksum(
      "Ethereum address fails EIP-55 checksum validation",
    ));
  }
  Ok(compact)
}
